### PR TITLE
core/thread: add getters for thread_t

### PR DESCRIPTION
**Pull request by kaspar030**
### Contribution description

This PR creates some getters for internal thread_t defines as used by ps().
Main reason is "cleanup", I also want to be able to have alternative core implementations.

### Testing procedure

Binaries should not change.

### CODE REVIEW

**Comment by kaspar030** [comment]
(squashed)

**Review by maribu — APPROVED** [review]
ACK. Please uncrustify before merge.

File: sys/ps/ps.c [hunk 1]
```diff
@@ -115,20 +115,20 @@ void ps(void)
...
 #ifdef CONFIG_THREAD_NAMES
-                   p->name,
+                   thread_get_name(p),
 #endif
                    sname, queued, p->priority
 #ifdef DEVELHELP
-                   , p->stack_size, stacksz, stack_free,
-                   (void *)p->stack_start, (void *)p->sp
+                   , (uint32_t)thread_get_stacksize(p), stacksz, stack_free,
+                   (void *)thread_get_stackstart(p), (void *)thread_get_sp(p)
```
Comment (benpicco) [inline sys/ps/ps.c:131]: You can now drop the cast

File: core/include/thread.h [hunk 2]
```diff
@@ -526,6 +526,81 @@ static inline bool thread_is_active(const thread_t *thread)
...
+
+/**
+ * Get stored Stack Pointer of thread.
+ *
+ * *Only provides meaningful value if the thread is not currently running!*.
+ *
+ * @param   thread thread to work on
+ * @returns current stack pointer
+ */
+static inline void *thread_get_sp(const thread_t *thread)
```
Comment (benpicco) [inline core/include/thread.h:554]: I wonder if those should return `const void *` so people don't get funky ideas

Comment (kaspar030) [reply core/include/thread.h:554]: should it? I don't mind, it could mean casting down the line (e.g., in ps()).

Comment (chrysn) [reply core/include/thread.h:554]: Where's the "grinning smiley with devil's horns" react emoji when you need it? >;-D

Comment (benpicco) [reply core/include/thread.h:554]: Does `ps` require non-const access to the thread stack? :fearful:

Comment (kaspar030) [reply core/include/thread.h:554]: `((void *)thread_get_sp(thread))[123] = 0xD1ED1E;`

ah wait, stack grows downwards. :/

Comment (chrysn) [reply core/include/thread.h:554]: If the thread is reliably stopped, couldn't it even be used meaningfully by resetting the thread's high water mark? (Then the non-const would be justified, although the constness of a pointer that arbitrary arithmetic is later done to has little meaning anyway).

Comment (kaspar030) [reply core/include/thread.h:554]: Like, for resetting the stack measurement? That could be useful.

Do we need const protection from devs that want to intentionally abuse a given stack pointer? `int foo; void *sp = &foo;` goes a long way for those inclined.

Comment (benpicco) [reply core/include/thread.h:554]: If you can think of use cases where `void *` is needed than that's fine.
I just try to use `const` whenever possible.

**Comment by benpicco** [comment]
Feel free to squash!

File: sys/ps/ps.c [hunk 3]
```diff
@@ -115,20 +115,20 @@ void ps(void)
...
                    " | %2d.%03d%% |  %8u  | %10"PRIu32" "
 #endif
                    "\n",
-                   p->pid,
+                   thread_get_pid(p),
 #ifdef CONFIG_THREAD_NAMES
-                   p->name,
+                   thread_get_name(p),
 #endif
                    sname, queued, p->priority
```
Comment (benpicco) [inline sys/ps/ps.c:128]: Not
```suggestion
                   sname, queued, thread_get_priority(p)
```
?

Comment (kaspar030) [reply sys/ps/ps.c]: yes, thanks!

**Comment by kaspar030** [comment]
I added previously forgotten `thread_get_priority()` and thanks to @benpicco raven eye also made ps() use that.

File: core/include/thread.h [hunk 4]
```diff
@@ -526,6 +537,81 @@ static inline bool thread_is_active(const thread_t *thread)
...
+
+/**
+ * Get PID of thread.
+ *
+ * This is a simple getter for thread->pid.
+ *
+ * @param   thread thread to work on
+ * @returns thread pid
+ */
+static inline kernel_pid_t thread_get_pid(const thread_t *thread)
```
Comment (benpicco) [inline core/include/thread.h:594]: It's a bit awkward that we now have [`thread_getpid()`](https://doc.riot-os.org/group__core__thread.html#gab68e945fba9216126e255648a9a4ee8a) as well as `thread_get_pid()` 🏌️

File: core/include/thread.h [hunk 5]
```diff
@@ -526,6 +537,81 @@ static inline bool thread_is_active(const thread_t *thread)
...
+{
+    return thread->pid;
+}
```
Comment (benpicco) [inline core/include/thread.h:597]: How about

```suggestion
static inline kernel_pid_t thread_get_pid(const thread_t *thread)
{
    extern volatile kernel_pid_t sched_active_pid;

    if (thread) {
        return thread->pid;
    } else {
        return sched_active_pid;
    }
}
```

Comment (kaspar030) [reply core/include/thread.h:597]: hm, not sure. the dual name is an issue, but I'd rather not change "thread_getpid()". (that would be the reason for overloading thread_get_pid(), right?)

Comment (maribu) [reply core/include/thread.h:597]: Maybe name it `thread_getpid_of()`? That name would make it more obvious which of the two requires an extra argument.

**Comment by kaspar030** [comment]
sorry for all the force-pushing.

**Comment by kaspar030** [comment]
I think this is good to go!

**Review by benpicco — APPROVED** [review]

**Comment by kaspar030** [comment]
Thanks everyone!

**Pull request merged**